File: Chapter1/JavaScript/question-1.1.js
// Is unique string.

// Without additional datastructures.
// Runtime - O(n^2)
const isUnique = (value) => {
  if (value && value.length) {
    for (let i = 0; i < value.length; i++) {
      const item = value[i];
      for (let j = i + 1; j < value.length; j++) {
        if (value[j] === item) {
          return false;
        }
      }
    }
  }
  return true;
}

// With datastruct.
// Runtime - O(n)
const isUniqueWithHashTable = value => {
  if (value && value.length) {
    const strStore = new Map()
    for (let i = 0; i < value.length; i++) {
      const item = value[i];
      if (item && strStore.get(item)) {
        return false;
      }
      strStore.set(item, true)
    }
  }
  return true;
}

// Tests
console.log('is unqiue 123456a', isUnique('123456a'))
console.log('is unqiue 12233456a', isUnique('12233456a'))
console.log('is unqiue 123456a', isUniqueWithHashTable('123456a'))
console.log('is unqiue 1223456a', isUniqueWithHashTable('1223456a'))

File: Chapter1/Python/question-1.1.py
# Is unique function, does not use extra datastructures.
def is_unique(value):
  if (value and len(value)):
    i = 0
    for item in value:
      for j in range(i + 1, len(value)):
        if (value[j] == item):
          return False
      i = i + 1
  return True

# Is unique function, uses hashtable.
def is_unique_with_hashtable(value):
  if (value and len(value)):
    str_store = { }
    for item in value:
      stringItem = str(item)
      if (item and item in str_store):
        return False
      str_store[stringItem] = True
  return True

print('is unqiue 123456a', is_unique('123456a'))
print('is unqiue with hashtable 123456a', is_unique_with_hashtable('123456a'))